/**
 * Generated by [[bio.ferlab.datalake.spark3.utils.ClassGenerator]]
 * on 2021-02-25T11:04:32.022
 */
package bio.ferlab.datalake.testutils.models.raw

import bio.ferlab.datalake.testutils.models.normalized.{AminoAcids, CODONS, Exon, Intron}

case class RawVcf(`contigName`: String = "chr1",
                  `start`: Long = 69896,
                  `end`: Long = 69897,
                  `names`: List[String] = List("rs200676709"),
                  `referenceAllele`: String = "T",
                  `alternateAlleles`: List[String] = List("C"),
                  `qual`: Double = 2324.17,
                  `splitFromMultiAllelic`: Boolean = false,
                  `INFO_FILTERS`: List[String] = List("PASS"),
                  `INFO_AC`: List[Int] = List(59),
                  `INFO_SOR`: Double = 0.005,
                  `INFO_ReadPosRankSum`: Double = 0.736,
                  `INFO_AN`: Int = 88,
                  `INFO_AF`: List[Double] = List(0.67),
                  `INFO_FS`: Double = 57.081,
                  `INFO_LOD`: Option[Double] = None,
                  `INFO_DP`: Int = 330,
                  `INFO_MQ`: Double = 15.54,
                  `INFO_QD`: Double = 8.07,
                  `INFO_END`: Option[Int] = None,
                  `INFO_CSQ`: List[InfoCSQ] = List(InfoCSQ()),
                  `INFO_DB`: Boolean = true,
                  `INFO_MQRankSum`: Double = 2.054,
                  `INFO_OLD_MULTIALLELIC`: Option[String] = None,
                  `genotypes`: List[SNV_GENOTYPES] = List(SNV_GENOTYPES(), SNV_GENOTYPES(`sampleId` = "22222", `calls` = List(0, 0)), SNV_GENOTYPES(`sampleId` = "33333")))


case class RawVcfWithInfoAnn(`contigName`: String = "chr1",
                  `start`: Long = 69896,
                  `end`: Long = 69897,
                  `names`: List[String] = List("rs200676709"),
                  `referenceAllele`: String = "T",
                  `alternateAlleles`: List[String] = List("C"),
                  `qual`: Double = 2324.17,
                  `splitFromMultiAllelic`: Boolean = false,
                  `INFO_FILTERS`: List[String] = List("PASS"),
                  `INFO_AC`: List[Int] = List(59),
                  `INFO_SOR`: Double = 0.005,
                  `INFO_ReadPosRankSum`: Double = 0.736,
                  `INFO_AN`: Int = 88,
                  `INFO_AF`: List[Double] = List(0.67),
                  `INFO_FS`: Double = 57.081,
                  `INFO_LOD`: Option[Double] = None,
                  `INFO_DP`: Int = 330,
                  `INFO_MQ`: Double = 15.54,
                  `INFO_QD`: Double = 8.07,
                  `INFO_END`: Option[Int] = None,
                  `INFO_ANN`: List[InfoCSQ] = List(InfoCSQ()),
                  `INFO_DB`: Boolean = true,
                  `INFO_MQRankSum`: Double = 2.054,
                  `INFO_OLD_MULTIALLELIC`: Option[String] = None,
                  `genotypes`: List[SNV_GENOTYPES] = List(SNV_GENOTYPES(), SNV_GENOTYPES(`sampleId` = "22222", `calls` = List(0, 0)), SNV_GENOTYPES(`sampleId` = "33333")))


case class InfoCSQ(`Allele`: String = "C",
                   `Consequence`: List[String] = List("synonymous_variant"),
                   `IMPACT`: String = "LOW",
                   `SYMBOL`: String = "OR4F5",
                   `Gene`: String = "ENSG00000186092",
                   `Feature_type`: String = "Transcript",
                   `Feature`: String = "ENST00000335137",
                   `BIOTYPE`: String = "protein_coding",
                   `EXON`: Exon = Exon("1", 1),
                   `INTRON`: Intron = Intron(None, None),
                   `HGVSc`: String = "ENST00000335137.4:c.807T>C",
                   `HGVSp`: String = "ENSP00000334393.3:p.Ser269%3D",
                   `cDNA_position`: Int = 843,
                   `CDS_position`: Int = 807,
                   `Protein_position`: Int = 269,
                   `Amino_acids`: AminoAcids = AminoAcids("S", None),
                   `Codons`: CODONS = CODONS("tcT", "tcC"),
                   `Existing_variation`: List[String] = List("rs200676709"),
                   `DISTANCE`: Option[Int] = None,
                   `STRAND`: Int = 1,
                   `FLAGS`: Option[List[String]] = None,
                   `PICK`: String = "1",
                   `VARIANT_CLASS`: String = "SNV",
                   `SYMBOL_SOURCE`: String = "HGNC",
                   `HGNC_ID`: String = "HGNC:14825",
                   `CANONICAL`: String = "YES",
                   `RefSeq`: String = "NM_001005484.1&NM_001005484.1&NM_001005484.2",
                   `HGVS_OFFSET`: Option[String] = None,
                   `HGVSg`: String = "chr1:g.69897T>C",
                   `CLIN_SIG`: Option[String] = None,
                   `SOMATIC`: Option[String] = None,
                   `PHENO`: Option[String] = None,
                   `PUBMED`: Option[String] = Some("29135816"),
                   `CADD_raw_rankscore`: Option[String] = None,
                   `DANN_rankscore`: Option[String] = None,
                   `Ensembl_geneid`: Option[String] = None,
                   `Ensembl_transcriptid`: Option[String] = None,
                   `ExAC_AC`: Option[String] = None,
                   `ExAC_AF`: Option[String] = None,
                   `FATHMM_converted_rankscore`: Option[String] = None,
                   `FATHMM_pred`: Option[String] = None,
                   `GTEx_V7_tissue`: Option[String] = None,
                   `Interpro_domain`: Option[String] = None,
                   `LRT_converted_rankscore`: Option[String] = None,
                   `LRT_pred`: Option[String] = None,
                   `Polyphen2_HVAR_pred`: Option[String] = None,
                   `Polyphen2_HVAR_rankscore`: Option[String] = None,
                   `REVEL_rankscore`: Option[String] = None,
                   `SIFT_converted_rankscore`: Option[String] = None,
                   `SIFT_pred`: Option[String] = None,
                   `UK10K_AC`: Option[String] = None,
                   `UK10K_AF`: Option[String] = None,
                   `clinvar_MedGen_id`: Option[String] = None,
                   `clinvar_OMIM_id`: Option[String] = None,
                   `clinvar_Orphanet_id`: Option[String] = None,
                   `clinvar_clnsig`: Option[String] = None,
                   `clinvar_id`: Option[String] = None,
                   `clinvar_trait`: Option[String] = None,
                   `gnomAD_exomes_AC`: Option[String] = None,
                   `gnomAD_exomes_AF`: Option[String] = None,
                   `gnomAD_genomes_AC`: Option[String] = None,
                   `gnomAD_genomes_AF`: Option[String] = None,
                   `phyloP17way_primate_rankscore`: Option[String] = None,
                   `rs_dbSNP151`: Option[String] = None)


case class SNV_GENOTYPES(`sampleId`: String = "11111",
                         `conditionalQuality`: Int = 30,
                         `F1R2`: List[Int] = List(0, 0),
                         `PRI`: Option[List[Double]] = None,
                         `filters`: List[String] = List("PASS"),
                         `posteriorProbabilities`: List[Double] = List(7.795, 3.805, 3.795),
                         `SB`: Option[List[Int]] = None,
                         `F2R1`: List[Int] = List(0, 1),
                         `alleleDepths`: List[Int] = List(0, 30),
                         `ICNT`: Option[List[Int]] = None,
                         `AF`: List[Double] = List(1.0),
                         `phased`: Boolean = false,
                         `calls`: List[Int] = List(0, 1),
                         `MIN_DP`: Option[Int] = None,
                         `phredLikelihoods`: List[Int] = List(42, 3, 0),
                         `LOD`: Option[Double] = None,
                         `depth`: Int = 1,
                         `SPL`: Option[List[Int]] = None,
                         `PS`: Option[Int] = None,
                         `MB`: Option[List[Int]] = None)
